src/MetadataManager/Metadata.ts

```diff
@@ -0,0 +1,207 @@
+  }
+
+  public static async getEntry(uri: vscode.Uri, hash: string) {
+    const metaObj = await Metadata.getObj(hash);
+    const relPath = vscode.workspace.asRelativePath(uri);
+    if (metaObj && metaObj[relPath] && Object.keys(metaObj[relPath]).length !== 0) {
+      return metaObj[relPath];
+    }
+    // TODO: Error Handling
+    return undefined;
+  }
+
+  public static async setEntry(uri: vscode.Uri, hash: string, entry: any) {
+    const workspaceroot = obtainWorkspaceRoot();
+    const jsonUri = vscode.Uri.joinPath(
+        vscode.Uri.file(workspaceroot),
+        `.one-vscode/info/hash_objects/${hash.substring(0, 2)}/${hash.substring(2)}.json`);
+
+    const relPath = vscode.workspace.asRelativePath(uri);
+    const metaObj = await Metadata.getObj(hash);
+    if (!metaObj[relPath]) {
+      metaObj[relPath] = {};
+    }
+    for (let key in entry) {
+      metaObj[relPath][key] = entry[key];
+    }
+    await vscode.workspace.fs.writeFile(
+        jsonUri, Buffer.from(JSON.stringify(metaObj, null, 4), 'utf8'));
+  }
+
+  public static async createDefault(uri: vscode.Uri, hash: string) {
+    if (!vscode.workspace.workspaceFolders) {
+      return;
+    }
+
+    let metaObj: any = await Metadata.getObj(hash);
+    const relPath = vscode.workspace.asRelativePath(uri);
+
+    if (metaObj === undefined) {
+      metaObj = {};
+      await Metadata.setObj(hash, metaObj);
+    }
+    if (metaObj[relPath] === undefined) {
+      metaObj[relPath] = {};
+    }
+
+    const filename: any = relPath.split('/').pop();
+    const stats: any = fs.statSync(uri.fsPath);
+
+    metaObj[relPath]['name'] = filename;
+    metaObj[relPath]['file-extension'] = filename.split('.').pop();
+    metaObj[relPath]['create-time'] = stats.birthtime;
+    metaObj[relPath]['modified-time'] = stats.mtime;
+    metaObj[relPath]['is-deleted'] = false;
+    await Metadata.setObj(hash, metaObj);
+  }
+
+  // NOTE When the deleted file is recovered, metadata is also recovered.
+  // For that situation, metadata is not deleted but deactivated.
+  public static async disable(uri: vscode.Uri, hash: string) {
+    if (!isOneExplorerTargetFile(uri)) {
+      return;
+    }
+    // step 1. Find hash object with hash value
+    const metaObj = await Metadata.getObj(hash);
+
+    // step 2. Check if the hash object has the deleted uri
+    const relPath = vscode.workspace.asRelativePath(uri);
+    if (metaObj === undefined || metaObj[relPath] === undefined ||
+        Object.keys(metaObj[relPath]).length === 0) {
+      return;
+    }
+
+    // step 3. deactivate (set 'is_deleted') that path.
+    metaObj[relPath]['is-deleted'] = true;
+    await Metadata.setObj(hash, metaObj);
+  }
+
+  public static async delete(uri: vscode.Uri, hash: string) {
+    const metaObj = await Metadata.getObj(hash);
+    const relPath = vscode.workspace.asRelativePath(uri);
+    delete metaObj[relPath];
```

Comment:
I wonder what happens when metaObj[relPath] not exists. If it throws, it's critical. Could you make a unit tests for that, maybe after this PR?

Reply:
When metaObj[relPath] does not exist, if you try to delete, nothing happens.
[Reference here](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Operators/delete#%EC%84%A4%EB%AA%85)
After this PR, I will add this case to the unit test just in case.